engine/execution/ingestion/engine.go

```diff
@@ -440,6 +453,43 @@ func (e *Engine) BlockProcessable(b *flow.Header) {
 	}
 }
 
+// BlockFinalized implements part of state.protocol.Consumer interface.
+// Method gets called for every finalized block
+func (e *Engine) BlockFinalized(h *flow.Header) {
+
+	if e.pauseExecution {
+		return
+	}
+
+	stopEnabled, stopHeight, stopCrash := e.stopAtHeight.Get()
+
+	// skip if no stop at height has been requested
+	if !stopEnabled {
+		return
+	}
+
+	// once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// if this block's parent has been executed, we are safe to stop or crash.
+	// This will happen during normal execution, where blocks are executed before they are finalized
+	// However, if the node is not up-to-date, finalization might lead, but we want to crash only after
+	// the execution reached the stop height
+	if h.Height == stopHeight {
+
+		executed, err := state.IsBlockExecuted(e.unit.Ctx(), e.execState, h.ParentID)
+		if err != nil {
+			e.log.Error().Err(err).Str("block_id", h.ID().String()).Msg("failed to check if the block has been executed")
+			return
+		}
+
+		if executed {
+			e.stopExecution(stopCrash, stopHeight)
+		} else {
+			e.stopAfterExecuting = h.ParentID
+		}
```

Comment:
I'm little concerned about the race condition here. 

`stopAfterExecuting` is checked by block execution. However block execution happens concurrently with block finalization, therefore there will be race condition here. 

Consider this case:

Let's say stop height is 100.
The two events: block 100 is finalized and block 99 is executed happen at the same time. 
Here, before `e.stopAfterExecuting = h.ParentID` in `BlockFinalized` is executed, the check `if e.stopAfterExecuting == executableBlock.ID() {` in `executeBlock ` still return `false`, so it won't crash the node. However, it just missed the last change to crash the EN. After `e.stopAfterExecuting = h.ParentID` is executed, the `e.stopAfterExecuting` is now block 99's ID, but since 99 has been executed, the check `if e.stopAfterExecuting == executableBlock.ID() ` will never return true anymore. EN will never crash.

Right now `stopHeight`, `stopEnabled`, `stopAfterExecuting`, `stopCrash` are read and written by different logic concurrently, I think we might need some synchronization here.

Reply:
Hmmm, ok, so the case when 99 has already been executed when 100 is finalized is covered in line 478.
The case where 99 has not finished execution before 100 is finalized, is covered
100 will never be executed, because we skip based on height.

So the problem is when finalization of 100 happens precisely before execution of 99 finishes but after it has started executing.

I think best approach would be to simply make sure both functions don't run at the same time - do you think simply using mutex to sync `OnBlockFinalized`  and final part of `OnBlockExecuted` when then check happens? 
Since its once per block and finalization handler is rather small I doubt there would be a lot of contention. Agree?

Regarding synchronization - Yes, I was considering that, but `stopEnabled`, `stopHeight`, `stopCrash` are synchronized, and `stopAfterExecuting` and `pauseExecution` are write-once really, so I don't think there would be a race there

Reply:
I think one simple solution to solve the race condition is to check twice. 

```
	// we want to stop at stopHeight, meaning block at stopHeight will not be executed.
	lastHeightToExecute := stopHeight - 1
	if h.Height == lastHeightToExecute {
	        
			executed, err := checkAndStopExecution(stopCrash, lastHeightToExecute)

		        // there is a race condition in checkAndStopExecution, after setting stopAfterExecuting,
		        // the block might just have been executed, meaning the previous check was stale. 
		        // we can avoid this race condition by checking again.

			checkExecutedAgain, err := checkAndStopExecution(stopCrash, lastHeightToExecute)
	}

```

Reply:
Yes, this seems to be great fix.
I have added a test to confirm that!